### PR TITLE
Add dynamic shape support in TensorRT execution provider

### PR DESCRIPTION
**Pull request by stevenlix**
Since TensorRT6.0 released a couple of months ago, we identified/fixed many issues in ONNX-TensorRT parser and TensorRT itself with Nvidia team. With those fixes, now we can run full dynamic shape models such as BERT and Faster/Mask-rcnn in OnnxRuntime-TensorRT. 
This PR adopts latest parser and TensorRT6.0 APIs, and enables dynamic shape support in TensorRT execution provider.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
Is it possible to sqash it a bit, CF only supports up to 64 commits in a PR. Hard to see everything even on GH.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 1]
```diff
@@ -317,30 +404,67 @@ TensorrtExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
...
+
   // Construct subgraph capability from node list
   std::vector<std::unique_ptr<ComputeCapability>> result;
   int counter = 0;
-  for (const auto& group : supported_nodes_vector) {
+  for (const auto& group : post_processed_supported_nodes_vector) {
     if (!group.first.empty()) {
       std::unique_ptr<IndexedSubGraph> sub_graph = GetSubGraph(group, counter, graph);
-      result.push_back(onnxruntime::make_unique<ComputeCapability>(std::move(sub_graph)));
+      result.push_back(std::make_unique<ComputeCapability>(std::move(sub_graph)));
```
Comment (HectorSVC) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: I think we should use onnxruntime::make_unique instead of std::make_unique

Comment (stevenlix) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc]: good catch. Will change

**Comment by jywu-msft** [comment]
Add some tests for dynamic shapes?
Also some documentation on how to run some of the new models - with some mention of some of the configurations and potential need to run symbolic inference script.
Something like https://github.com/microsoft/onnxruntime/blob/master/docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb would be really nice

**Comment by jywu-msft** [comment]
looks like there's a unit test failure on windows?
the test segfaults.

1: [ RUN      ] TensorOpTest.ReshapeWithEmptyDim

File: BUILD.md [hunk 2]
```diff
@@ -189,14 +189,14 @@ See more information on the TensorRT Execution Provider [here](./docs/execution_
...
-   * The TensorRT execution provider for ONNX Runtime is built and tested with TensorRT 6.0.1.5 but validated with the feature set equivalent to TensorRT 5. Some TensorRT 6 new features such as dynamic shape is not available at this time.
+   * The TensorRT execution provider for ONNX Runtime is built and tested with TensorRT 6.0.1.5.
    * The path to TensorRT installation must be provided via the `--tensorrt_home parameter`.
 
 #### Build Instructions
 ##### Linux
 
 ```
-./build.sh --cudnn_home <path to cuDNN e.g. /usr/lib/x86_64-linux-gnu/> --cuda_home <path to folder for CUDA e.g. /usr/local/cuda> --use_tensorrt --tensorrt_home <path to TensorRT home>
+./build.sh --use_full_protobuf --cudnn_home <path to cuDNN e.g. /usr/lib/x86_64-linux-gnu/> --cuda_home <path to folder for CUDA e.g. /usr/local/cuda> --use_tensorrt --tensorrt_home <path to TensorRT home>
```
[the remark this answers is not in the document]
Comment (jywu-msft) [reply BUILD.md]: i double checked and , yes, the build.py will light full_protobuf flag if use_tensorrt is option is set. 
so --use_full_protobuf is not needed.

File: docs/execution_providers/TensorRT-ExecutionProvider.md [hunk 3]
```diff
@@ -7,20 +7,35 @@ With the TensorRT execution provider, the ONNX Runtime delivers better inferenci
...
+The TensorRT execution provider needs to be registered with ONNX Runtime to enable in the inference session. 
 ```
 InferenceSession session_object{so};
 session_object.RegisterExecutionProvider(std::make_unique<::onnxruntime::TensorrtExecutionProvider>());
 status = session_object.Load(model_file_name);
 ```
 The C API details are [here](../C_API.md#c-api).
 
+#### Sample
+To run Faster R-CNN model on TensorRT execution provider,
```
Comment (jywu-msft) [inline docs/execution_providers/TensorRT-ExecutionProvider.md]: it's better to do a notebook tutorial to be consistent with other EPs. 
let's do it as a separate future PR.

File: onnxruntime/test/providers/tensorrt/tensorrt_basic_test.cc [hunk 4]
```diff
@@ -103,85 +103,6 @@ TEST(TensorrtExecutionProviderTest, FunctionTest) {
   VerifyOutputs(fetches, expected_dims_mul_m, expected_values_mul_m);
 }
 
-TEST(TensorrtExecutionProviderTest, DynamicShapeTest) {
```
Comment (jywu-msft) [inline onnxruntime/test/providers/tensorrt/tensorrt_basic_test.cc]: why is this removed?

Comment (stevenlix) [reply onnxruntime/test/providers/tensorrt/tensorrt_basic_test.cc]: It's not needed since we added dynamic shape model in the mode tests for both Linux and Windows.

File: docs/execution_providers/TensorRT-ExecutionProvider.md (hunk 3 above)
Comment (stevenlix) [reply docs/execution_providers/TensorRT-ExecutionProvider.md]: Will do

File: tools/ci_build/build.py [hunk 5]
```diff
@@ -610,6 +608,31 @@ def run_onnx_tests(build_dir, configs, onnx_test_data_dir, provider, enable_mult
...
+        cmd_base = ['-e', 'tensorrt', '-j', '1'] 
+
+        #onnx test
+        if os.path.exists(onnx_test_data_dir):
+          onnx_test_cmd = cmd_base + [onnx_test_data_dir]
+          run_subprocess([exe] + onnx_test_cmd, cwd=cwd)
+
+        # model test
+        # TensorRT can run most of the model tests, but only part of them is enabled here to save CI build time.
+        if config != 'Debug' and os.path.exists(model_dir):
```
[the remark this answers is not in the document]
Comment (stevenlix) [reply tools/ci_build/build.py]: Yes, I agree we need to increase the coverage rather than decrease. I will separate the time consuming model test from the regular CI build in the future PR, then we can enable most of the model tests on TRT.
